src/gravel/cephadm/cephadm.py

```diff
@@ -64,24 +66,50 @@ async def call(self, cmd: str) -> Tuple[str, str, int]:
     async def run_in_background(self, cmd: List[str]) -> None:
         pass
 
-    async def _tee(self, reader: asyncio.StreamReader) -> str:
+    async def _tee(self, reader: asyncio.StreamReader,
+                   cb: Optional[Callable[[str], None]] = None) -> str:
         collected = StringIO()
         async for line in reader:
             msg = line.decode("utf-8")
             collected.write(msg)
+            if cb:
+                cb(msg)
         return collected.getvalue()
 
     #
     # command wrappers
     #
 
-    async def bootstrap(self, addr: str) -> Tuple[str, str, int]:
+    async def bootstrap(self, addr: str,
+                        percentcb: Callable[[int], None]
+                        ) -> Tuple[str, str, int]:
 
         if not addr:
             raise CephadmError("address not specified")
 
+        msg_to_percent: List[Tuple[str, int]] = [
+            ("cluster fsid", 5),
+            ("pulling container image", 7),
+            ("extracting ceph user uid/gid from container image", 15),
+            ("creating mon", 20),
+            ("waiting for mon to start", 25),
+            ("wrote config", 40),
+            ("creating mgr", 45),
+            ("waiting for mgr to start", 50),
+            ("enabling cephadm module", 60),
+            ("setting orchestrator backend to cephadm", 70),
+            ("adding host", 80),
+            ("bootstrap complete", 100)
+        ]
```

Comment:
super fragile, seems like this would justify a unit test with various types of output from cephadm ..

Reply:
https://github.com/ceph/ceph/pull/40103

Reply:
oof. Glad it didn't change any of those, eh? :)